packages/in_app_purchase/lib/src/in_app_purchase_tizen_platform.dart

```diff
@@ -0,0 +1,185 @@
+      response = const ProductsListApiResult(
+          cPStatus: 'fail to receive response',
+          cPResult: 'fail to receive response',
+          checkValue: 'fail to receive response',
+          totalCount: 0,
+          itemDetails: <ItemDetails>[]);
+    }
+
+    List<SamsungCheckoutProductDetails> productDetailsList =
+        <SamsungCheckoutProductDetails>[];
+
+    productDetailsList = response.itemDetails
+        .map((ItemDetails productWrapper) =>
+            SamsungCheckoutProductDetails.fromProduct(productWrapper))
+        .toList();
+    final List<String> invalidMessage = <String>[];
+    switch (response.cPStatus) {
+      case '100000':
+        break;
+      default:
+        invalidMessage.add(response.toJson().toString());
+    }
+
+    final ProductDetailsResponse productDetailsResponse =
+        ProductDetailsResponse(
+      productDetails: productDetailsList,
+      notFoundIDs: invalidMessage,
+      error: exception == null
+          ? null
+          : IAPError(
+              source: kIAPSource,
+              code: exception.code,
+              message: exception.message ?? '',
+              details: exception.details),
+    );
+    return productDetailsResponse;
+  }
+
+  @override
+  Future<void> restorePurchases({
+    String? applicationUserName,
+  }) async {
+    List<GetUserPurchaseListAPIResult> responses;
+
+    responses = await Future.wait(<Future<GetUserPurchaseListAPIResult>>[
+      billingManager.requestPurchases(
+          applicationUserName, _identifiers.toList())
+    ]);
+
+    final List<PurchaseDetails> pastPurchases =
+        responses.expand((GetUserPurchaseListAPIResult response) {
+      return response.invoiceDetails;
+    }).map((InvoiceDetails purchaseWrapper) {
+      final SamsungCheckoutPurchaseDetails purchaseDetails =
+          SamsungCheckoutPurchaseDetails.fromPurchase(purchaseWrapper);
+
+      purchaseDetails.status = PurchaseStatus.restored;
+      return purchaseDetails;
+    }).toList();
+    _purchaseUpdatedController.add(pastPurchases);
+  }
+
+  @override
+  Future<bool> buyNonConsumable({required PurchaseParam purchaseParam}) async {
+    final BillingBuyData billingResultWrapper = await billingManager.buyItem(
+        appId: _appId,
+        serverType: _serverType,
+        orderItemId: purchaseParam.productDetails.id,
+        orderTitle: purchaseParam.productDetails.title,
+        orderTotal: purchaseParam.productDetails.price,
+        orderCurrencyId: purchaseParam.productDetails.currencyCode);
+
+    if (billingResultWrapper.payResult == 'SUCCESS') {
+      final String invoiceId =
+          billingResultWrapper.payDetails['InvoiceID'] ?? '';
+      final GetUserPurchaseListAPIResult responses = await billingManager
+          .requestPurchases(_identifiers.toList()[8], _identifiers.toList());
+
+      for (int i = 0; i < responses.invoiceDetails.length; i++) {
+        if (responses.invoiceDetails[i].invoiceID != invoiceId) {
```

Comment:
If I understand correctly,
```suggestion
        if (responses.invoiceDetails[i].invoiceID == invoiceId) {
```

Reply:
fixed.
[Use Future.then() in buyNonConsumable](https://github.com/flutter-tizen/plugins/pull/562/commits/ac657875a0678afdb13bf91f5bb01df8ff1e8075)